### DISCUSSION TITLE
Error on implementing periodic boundary condition

### DISCUSSION THREAD

**Discussion by avtarsinghh1991** [idaholab/moose#20232] [Q&A Modules: General] [answered]
Hello MOOSE Experts

I am working with a chemo-mechanical model in which displacement is diffusion induced.
For the sake of RVE, I want to implement periodic boundary conditions on the displacement field.

My domain is of square shape with coordinates (0,0), (1e-5,0), (1e-5,1e-5) and (0,1e-5). I have generated the mesh in cubit and also verified the domain size using paraview.

I am trying to put periodic boundary conditions on disp_x and disp_y for the left and right boundaries. But I am getting the following error:

```
*** ERROR ***
Periodic boundary neighbor not found

application called MPI_Abort(MPI_COMM_WORLD, 1) - process 3


*** ERROR ***
Periodic boundary neighbor not found
```

My code snippet for implementing periodic bc is as follows:

```
[./Periodic]
    [./left_right_periodic_disp_x]
      variable = disp_x
      primary = left1
      secondary = right1
      translation = '1e-05 0 0'
    [../]
    [./left_right_periodic_disp_y]
      variable = disp_y
      primary = left1
      secondary = right1
      translation = '1e-05 0 0'
    [../]
  [../]
```

Please let me know any suggestions you have?

Update: If I apply the periodic boundaries on top and bottom boundaries, I didn't get any error even if put the wrong translation value.
Thanks.

Best
Avtar

**Comment by ngrilli** [comment]
Dear @asingh-mit 

Since your domain is a square,
you can try with auto_direction

For instrance, see this one:
https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/SandfeldZaiser2015/SandfeldLoop.i

Best Regards,
Nicolò

**Reply by avtarsinghh1991** [reply]
Hey @ngrilli 
Thank you very much. But I am importing a mesh file not generating it using a mesh generator. It cannot be used in that scenario.
Please correct me if I am wrong.

Even though I have tried and get the following error.

```
*** ERROR ***
Task init_mesh is not registered to build MeshGenerator derived objects
```

Best
Avtar

**Reply by GiudGiud** [reply]
You can import the mesh using the FileMeshGenerator. 
Then you can use mesh generators to modify it. 
https://mooseframework.inl.gov/source/meshgenerators/FileMeshGenerator.html

**Reply by avtarsinghh1991** [reply]
Hi @GiudGiud 

Thank you very much.

I am trying a simple problem. I have a square domain of size 1x1 afterward I am scaling the mesh. In this simple problem, the periodic boundary condition is not working. Please let me know any issues with the input file.

```
[Mesh]
   [./initial_mesh]
     type =FileMeshGenerator
     file = 'sq1.e'
  []

  [./scale]
    type = TransformGenerator
    input = initial_mesh
    transform = SCALE
    vector_value ='40 40 0'
  []
[]

[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]

  [./dot]
    type = TimeDerivative
    variable = u
  [../]

[]

[BCs]
  [./u_left]
    type = DirichletBC
    variable = u
    boundary = left1
    value = 1
  [../]

  [./Periodic]
    [./x_direction]
      variable = u
      primary = left1
      secondary = right1
      translation = '40 0 0'
    [../]
  [../]
[]

[Executioner]
  type = Transient
  dt = 1
  num_steps = 20
  solve_type = NEWTON
[]

[Outputs]
  execute_on = 'timestep_end'
  file_base = out
  exodus = true
[]
```

Best
Avtar

**Reply by GiudGiud** [reply]
u has both a Dirichlet and a periodic BC on left1? Might as well just use two DirichletBCs

**Reply by avtarsinghh1991** [reply]
@GiudGiud 

This is just a simple example problem I am trying to solve. Let's say with the Dirichlet BC, the value enforced on left should be transferred to the right using periodic boundary conditions.

Left and right boundaries get similar values if I use moose inbuilt MESHGENERATOR rather than the file.

Update: I have even tried the EqualConstraint options but nothing is working for me.

Best
Avtar

**Reply by avtarsinghh1991** [reply]
Is there any way, so that I can use the auto direction option for periodic boundary conditions while still importing the mesh file?

**Reply by GiudGiud** [reply]
I highly doubt we support that. Please use a single boundary conditions per boundary per variable first, and see if you can make that work for your cases.

PeriodicBCs are a whole different beast than DirichletBC. They are pretty much directly interfacing with libmesh to modify the numerical system.

But you are saying it works with moose generated meshes? Just not your custom one?

**Reply by avtarsinghh1991** [reply]
Yes, It is working with the MOOSE-generated meshes.

Actually, I am working on a chemo-mechanical problem in which displacement is induced. In that case, I only want to apply periodic boundary conditions for displacements on left, right, top, and bottom sides. No Dirichlet boundary conditions are needed. I tried to implement PeriodicBC there but it was not working so I am trying to implement it for the much simpler problem.

Best
Avtar

**Comment by dschwen** [comment]
Looks like your imported mesh has boundaries where the nodes do not exactly line up. Easiest solution is to remesh, or use an integrated constraint to enforce periodicity.

**Reply by avtarsinghh1991** [reply]
Thank you very much @dschwen. I will try the suggestions you proposed. Actually, I am using the scaling option in CUBIT to scale the model. I think that is distorting the mesh somehow. 

Best
Avtar

**Comment by avtarsinghh1991** [comment, marked as answer]
Hello @GiudGiud and @dschwen 

Thank you very much for your valuable insights and help. I resolved the issue.
Actually, periodic boundary conditions can only be applied to SIDESETS.
I was trying to implement it to NODESETS.

This has not been documented in the given text: 
https://mooseframework.inl.gov/source/actions/AddPeriodicBCAction.html

Best
Avtar

**Reply by GiudGiud** [reply]
Hello

There should be another documentation page for periodic boundary conditions, this is the "Action" page. I cant find it so it may not exist. 
Thanks for reporting this i ll add it to an upcoming documentation PR

Guillaume

**Reply by hugary1995** [reply]
Is this conclusion correct?? I'm a little surprised by that.

**Reply by GiudGiud** [reply]
Well I double checked the missing docs and I didnt find a better page than the Action. It's not explicitly specified what's supported in that page either.

wrt to nodesets and sidesets, what kind of variable are you using?

**Reply by GiudGiud** [reply]
note that you can get sidesets from nodesets and vice-versa from MOOSE. See this page:
https://mooseframework.inl.gov/source/mesh/MooseMesh.html

**Reply by hugary1995** [reply]
OP was using PBC on a nodal variable.

**Reply by hugary1995** [reply]
I mean, if PBC on a node set doesn't work for a nodal variable, that sounds like a bug -- not something that should be documented.

**Reply by avtarsinghh1991** [reply]
Hello @hugary1995 and @GiudGiud 

I generated the mesh using Trelis. It's a simple square domain.

First, I generated the nodesets and tried to apply the PBC, it was not working.

Input file, I was using is as below:

```
[Mesh]
   [./initial_mesh]
     type =FileMeshGenerator
     file = 'sq1.e'
  []

  [./scale]
    type = TransformGenerator
    input = initial_mesh
    transform = SCALE
    vector_value ='40 40 0'
  []
[]

[Variables]
  [./u]
    order = FIRST
    family = LAGRANGE
  [../]
[]

[Kernels]
  [./diff]
    type = Diffusion
    variable = u
  [../]

  [./dot]
    type = TimeDerivative
    variable = u
  [../]

[]

[BCs]
  [./u_left]
    type = DirichletBC
    variable = u
    boundary = left1
    value = 1
  [../]

  [./Periodic]
    [./x_direction]
      variable = u
      primary = left1
      secondary = right1
      translation = '40 0 0'
    [../]
  [../]
[]

[Executioner]
  type = Transient
  dt = 1
  num_steps = 20
  solve_type = NEWTON
[]

[Outputs]
  execute_on = 'timestep_end'
  file_base = out
  exodus = true
[]
```

I am using this file, just to understand the working of PBCs.

After struggling for 3 days. I generated the side sets in TRELIS and tried to apply the PBC, and then it started working absolutely fine.

Afterward, I applied the PBC on the displacement field in my actual chemo-mechanical problem, and they started working nicely.

Best
Avtar

**Reply by GiudGiud** [reply]
It's not just a PBC, it's trying to get the DirichletBC applied on one side to be transported to the other side by the PBC. 
If nodal PBCs were broken I think we would know about it
